pkg/cloudprovider/providers/gce/gce.go

```diff
@@ -1663,6 +1783,129 @@ func (gce *GCECloud) ListClusters() ([]string, error) {
 	return result, nil
 }
 
+func (gce *GCECloud) ListClusters() ([]string, error) {
+	allClusters := []string{}
+
+	for _, zone := range gce.managedZones {
+		clusters, err := gce.listClustersInZone(zone)
+		if err != nil {
+			return nil, err
+		}
+		// TODO: Scoping?  Do we need to qualify the cluster name?
+		allClusters = append(allClusters, clusters...)
+	}
+
+	return allClusters, nil
+}
+
 func (gce *GCECloud) Master(clusterName string) (string, error) {
 	return "k8s-" + clusterName + "-master.internal", nil
 }
+
+type gceInstance struct {
+	Zone  string
+	Name  string
+	ID    uint64
+	Disks []*compute.AttachedDisk
+}
+
+type gceDisk struct {
+	Zone string
+	Name string
+	Kind string
+}
+
+func (gce *GCECloud) getInstancesByNames(names []string) ([]*gceInstance, error) {
+	instances := make(map[string]*gceInstance)
+
+	for _, name := range names {
+		name = canonicalizeInstanceName(name)
+		instances[name] = nil
+	}
+
+	for _, zone := range gce.managedZones {
+		var remaining []string
+		for name, instance := range instances {
+			if instance == nil {
+				remaining = append(remaining, name)
+			}
+		}
+
+		if len(remaining) == 0 {
+			break
+		}
+
+		listCall := gce.service.Instances.List(gce.projectID, zone)
+
+		// Add the filter for hosts
+		listCall = listCall.Filter("name eq (" + strings.Join(remaining, "|") + ")")
+
+		listCall = listCall.Fields("items(name,id,disks)")
+
+		res, err := listCall.Do()
+		if err != nil {
+			return nil, err
+		}
+
+		for _, i := range res.Items {
+			name := i.Name
+			instance := &gceInstance{
+				Zone:  zone,
+				Name:  name,
+				ID:    i.Id,
+				Disks: i.Disks,
+			}
+			instances[name] = instance
+		}
+	}
+
+	instanceArray := make([]*gceInstance, len(names))
+	for i, name := range names {
+		instance := instances[name]
+		if instance == nil {
+			return nil, fmt.Errorf("failed to retrieve instance: %q", name)
+		}
+		instanceArray[i] = instances[name]
+	}
+
+	return instanceArray, nil
+}
+
+func (gce *GCECloud) getInstanceByName(name string) (*gceInstance, error) {
+	// Avoid changing behaviour when not managing multiple zones
+	if len(gce.managedZones) == 1 {
+		name = canonicalizeInstanceName(name)
+		zone := gce.managedZones[0]
+		res, err := gce.service.Instances.Get(gce.projectID, zone, name).Do()
+		if err != nil {
+			if !isHTTPErrorCode(err, http.StatusNotFound) {
```

Comment:
If `isHTTPErrorCode(err, http.StatusNotFound)` is true, line 1886 below will try to access `res.Zone`, causing nil-pointer dereference.